### PR TITLE
expr: optimize `to_char` for static format

### LINKED ISSUE #29068: Optimize `to_char(timestamp, format)` for static `format`s
Currently, when the `to_char` function is used in a query to convert a timestamp to a string, we compile the given format string into a `DateTimeFormat` for each processed datum. If there are many datums, this can have a noticeable performance impact. The format string is usually static, so it's worthwhile to apply an optimization where we only compile it once into a `DateTimeFormat` and then use that for each datum.

[Relevant Slack thread.](https://materializeinc.slack.com/archives/C0761MZ3QD9/p1723739253961039)

### PR DESCRIPTION
**Pull request by teskje**
This PR introduces unary `ToCharTimestamp` and `ToCharTimestampTz` variants that differ from their binary counterparts in that they have the format string already precompiled into a `DateTimeFormat`. This lets us skip the per-datum compilation in cases where the format string is statically known.

### Motivation

  * This PR adds a known-desirable feature.

Closes #29068 

### Tips for reviewer

I'm pretty unfamiliar with the code I touched here. Definitely check my work!

Most of the diff is adding protobuf support for `DateTimeFormat`.

Should we do more to test this? AFAICT most/all of the `to_char` uses in tests already supply a static format, so it seems that test coverage should already be good for this change.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by benesch** [comment]
Nice, excited about this!

**Review by ggevay — CHANGES_REQUESTED** [review]
The code looks perfect to me, but I think some more tests are needed. I can see `to_char` being called in 2 slts, but unfortunately none of them tests the new code path:

`dates-times.slt`: All the calls have _all_ of their arguments as literals (not just the format string). This means that we'll const-fold them [here](https://github.com/MaterializeInc/materialize/blob/54752df3c92f7e13c144b35e5f50740f9407a61d/src/expr/src/scalar.rs#L812) without invoking the new code.

`postgres/timestamp.slt` doesn't have the above issue, but [we don't run it at all](https://github.com/MaterializeInc/materialize/blob/6703ee22ab6c69e46b092302a1ec81e468c10b6b/test/sqllogictest/mzcompose.py#L530-L535)! (Note that there is another `timestamp.slt`, outside the `postgres` dir, which we do run, but doesn't have `to_char` calls.) When I try to manually run it locally, it immediately fails with
```
--- test/sqllogictest/postgres/timestamp.slt
    INSERT INTO t VALUES ('Mon Feb 10 17:32:01 1997 PST')
    Bail:test/sqllogictest/postgres/timestamp.slt:26 PlanFailure:test/sqllogictest/postgres/timestamp.slt:
    db error: ERROR: invalid input syntax for type timestamp: unknown units Feb: "Mon Feb 10 17:32:01 1997
    ----
    FAIL: bail=1 success=1 total=2
FAIL: bail=1 success=1 total=2
```
Is this known, @MaterializeInc/testing ? Were we running this at some point, and then we somehow regressed both in whether we run it and then in the functionality that it is testing? I guess we should get this slt working, right? Btw. maybe we should move those Postgres/Cockroach tests that we don't run into a `not_run` dir, to avoid people falsely thinking that they provide test coverage.

In addition to the non-const-folding tests that we can harvest from `postgres/timestamp.slt`, we should test a few more things:
- Call `EXPLAIN` to check that the new code is invoked. (Note that this should come in an slt that is not under the `postgres` dir even if we get `postgres/timestamp.slt` working, in order to limit the slts that one needs to rewrite when an EXPLAIN output format change happens.)
- null arguments: only the first, only the second, both. (With not both literals, to avoid const folding.) It's good to test this case, because null argument handling in `MirScalarExpr::reduce` is a bit brittle at the moment: the new code (as well as many other similar things) would panic when seeing a literal null argument, and the only thing which prevents this is the blanket `if (expr1.is_literal_null() || expr2.is_literal_null())` handling this case at the beginning of the match arm. And, if you feel like it, you could modify the new code to explicitly handle this case right there to make the code more robust against regressions if the blanket null handling changes for some reason. (And literal errors also have the same brittleness.)

The Nightly failures seem unrelated:
- The RQG one is known: #27954. (We should add an ignore for it in CI, to make it not cause flakes.)
- The Testdrive one might be a new issue, but seems unrelated. @MaterializeInc/testing, could you please check?

File: src/expr/src/scalar.rs [hunk 1]
```diff
@@ -1027,6 +1028,24 @@ impl MirScalarExpr {
...
+                        } else if *func == BinaryFunc::ToCharTimestamp && expr2.is_literal() {
+                            let format_str = expr2.as_literal_str().unwrap();
+                            *e = MirScalarExpr::CallUnary {
+                                func: UnaryFunc::ToCharTimestamp(func::ToCharTimestamp {
+                                    format_string: format_str.to_string(),
+                                    format: DateTimeFormat::compile(format_str),
+                                }),
+                                expr: Box::new(expr1.take()),
+                            };
+                        } else if *func == BinaryFunc::ToCharTimestampTz && expr2.is_literal() {
```
Comment (sdht0) [inline src/expr/src/scalar.rs:1040]: Can these two `if` blocks be folded into one and do the conditional when constructing `func` instead?

File: src/expr/src/scalar/func/impls/timestamp.rs [hunk 2]
```diff
@@ -721,3 +722,53 @@ impl fmt::Display for TimezoneTimestampTz {
...
+    }
+
+    fn output_type(&self, input: ColumnType) -> ColumnType {
+        ScalarType::String.nullable(input.nullable)
+    }
+}
+
+impl fmt::Display for ToCharTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "tocharts[{}]", self.format_string)
```
Comment (sdht0) [inline src/expr/src/scalar/func/impls/timestamp.rs:747]: Nit: use full name `ToCharTimestamp[{}]`. Also in `ToCharTimestampTz`.

**Review by sdht0 — COMMENTED** [review]
Nice work implementing all the proto changes. I took a look at this briefly on Friday but didn't manage to go that far.

Leaving a couple of suggestions after a cursory look at the diff.

**Comment by nrainer-materialize** [comment]
> Is this known, @MaterializeInc/testing ? Were we running this at some point, and then we somehow regressed both in whether we run it and then in the functionality that it is testing? I guess we should get this slt working, right? Btw. maybe we should move those Postgres/Cockroach tests that we don't run into a `not_run` dir, to avoid people falsely thinking that they provide test coverage.

I think that these test files are run. However, they are run with `--no-fail`:
https://github.com/MaterializeInc/materialize/blob/6073598ef917b5d5870a54ca42e92ec402bb706b/test/sqllogictest/mzcompose.py#L551-L556

The compute team is or at least used to be the code owner of the SLT tests. Happy to support with any changes. (Though, I don't know if we want and are allowed to changes these files copied from somewhere else.)

File: src/expr/src/scalar/func/impls/timestamp.rs (hunk 2 above)
Comment (teskje) [reply src/expr/src/scalar/func/impls/timestamp.rs:747]: This is the same name used for the `BinaryFunc` and I'd like to keep them consistent. We could of course change the name of the `BinaryFunc` too, but that would introduce an inconsistency with other `BinaryFunc`s which also use this style of abbreviated names.

Comment (ggevay) [reply src/expr/src/scalar/func/impls/timestamp.rs:747]: Yes, a large number of functions currently have the same issue. I think at some point we'll want to make the function names printed in EXPLAIN more easily understandable for users (i.e., closer to SQL), but that's out of scope for this PR.

File: src/expr/src/scalar.rs (hunk 1 above)
Comment (sdht0) [reply src/expr/src/scalar.rs:1040]: Yeah it's a trade-off. It just bugged me a little that most of the block is the same except the new `func`. But it's fine to duplicate it for now to keep it consistent with the existing pattern, as you mentioned.

Just for reference, I had the following in mind:
```rust
else if matches!(
    *func,
    BinaryFunc::ToCharTimestamp | BinaryFunc::ToCharTimestampTz
) && expr2.is_literal()
{
    let fmt_str = expr2.as_literal_str().unwrap();
    let fmt = func::format::DateTimeFormat::compile(&fmt_str);
    *e = MirScalarExpr::CallUnary {
        func: if *func == BinaryFunc::ToCharTimestamp {
            UnaryFunc::ToCharTimestamp(fmt)
        } else {
            UnaryFunc::ToCharTimestampTz(fmt)
        },
        expr: Box::new(expr2.take()),
    };
}
```

File: src/expr/src/scalar/func/impls/timestamp.rs (hunk 2 above)
Comment (sdht0) [reply src/expr/src/scalar/func/impls/timestamp.rs:747]: We could possibly start by changing the string for just this function. But resolving this for now.

**Comment by teskje** [comment]
Thanks for the reviews! I added a second commit with a couple tests. They only test the static-format-optimized version of `to_char`, not the regular one since the PR doesn't touch that. If you insist I can also add tests for the dynamic one, but that feels slightly out of scope here.

I'm pretty sure the `postgres/` tests are mostly not expected to work, which is why we ignore failures in them. In the case of `timestamp.slt` there are a lot of formats we don't support at all in Mz, and starting to support them would be a major effort I think. Definitely not something that fits in here.

> Note that this should come in an slt that is not under the `postgres` dir even if we get `postgres/timestamp.slt` working, in order to limit the slts that one needs to rewrite when an EXPLAIN output format change happens.

I've not put them into the slt files that have the other new `to_char` tests, which I suspect you might not like, based on this comment. I didn't see any slt files that looked like they were specifically for testing EXPLAIN output, so I'm not sure where else to put them. Also I don't quite follow the concerns about having many slts that need to be rewritten. With `--rewrite-results` that seems like a non-issue to me?

> And, if you feel like it, you could modify the new code to explicitly handle this case right there to make the code more robust against regressions if the blanket null handling changes for some reason.

I did not feel like it! 😬 If we break the null handling we will hopefully notice based on tests failing. Certainly the new ones added in this PR will fail.

File: test/sqllogictest/timestamp.slt [hunk 3]
```diff
@@ -13,9 +13,6 @@ ALTER SYSTEM SET enable_binary_date_bin = true
 ----
 COMPLETE 0
 
-# Tests adapted from Postgresql
-# https://github.com/postgres/postgres/blob/master/src/test/regress/sql/timestamp.sql
-
```
Comment (teskje) [inline test/sqllogictest/timestamp.slt:18]: I removed this comment. The current version of the file behind that link looks very different from what's in here, so mentioning it here is more confusing than helpful.

**Comment by ggevay** [comment]
Thanks for adding the tests!

> I'm pretty sure the `postgres/` tests are mostly not expected to work, which is why we ignore failures in them. In the case of `timestamp.slt` there are a lot of formats we don't support at all in Mz, and starting to support them would be a major effort I think. Definitely not something that fits in here.

Ok, yes, this is out of scope for this PR. (I still think it would be great to at least move these into a separate dir. Maybe I'll open a PR for that later.)

> > Note that this should come in an slt that is not under the `postgres` dir even if we get `postgres/timestamp.slt` working, in order to limit the slts that one needs to rewrite when an EXPLAIN output format change happens.

> I've not put them into the slt files that have the other new `to_char` tests, which I suspect you might not like, based on this comment. I didn't see any slt files that looked like they were specifically for testing EXPLAIN output, so I'm not sure where else to put them. Also I don't quite follow the concerns about having many slts that need to be rewritten. With `--rewrite-results` that seems like a non-issue to me?

You mean you did put them in the same slt files that have the other new `to_char` tests, right? This is fine! I just want to avoid EXPLAINs in exotic places, such as under `postgres` or `cockroach` or anything that is not run as part of the normal CI. But most files directly under `sqllogictest` are fine.

> > And, if you feel like it, you could modify the new code to explicitly handle this case right there to make the code more robust against regressions if the blanket null handling changes for some reason.

> I did not feel like it! 😬 If we break the null handling we will hopefully notice based on tests failing. Certainly the new ones added in this PR will fail.

Ok, that's fine!

**Review by ggevay — APPROVED** [review]

**Pull request merged**